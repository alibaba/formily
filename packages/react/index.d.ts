type ParamsWithContext<P extends any = any, C extends any = any> = (
  props: React.PropsWithChildren<P>,
  context: React.Context<C>
) => JSX.Element;

type RegistedFormFieldsDescription = {
  [type: string]: any;
}

type FieldConnectOptions = {
  valueName?: string;
  eventName?: string;
  getValueFromEvent?: any;
  getProps?: (schemaProps: object, xProps: object) => any;
  getComponent?: (oldComponent: any, schemaProps: object, xProps: object) => any;
}

export type SchemaFormProps = {
  className?: string;
  [x: string]: any;
};


export type SchemaFieldProps = {
  [x: string]: any;
};


export const SchemaForm: React.SFC<SchemaFormProps>;


export const Field: ParamsWithContext<SchemaFieldProps>;

/**
 * set the validation internationalization
 */
export const setValidationLocale: (locale: Record<string, any>) => void;

/**
 * set the validation langage
 */
export const setValidationLanguage: (lang: string) => void;

declare type FormActionType = {
  getFormState: (callback?: any) => any;
  getFieldState: (path?: string, callback: any) => any;
  setFormState: (callback?: any) => any;
  setFieldState: (path?: string, buffer?: any, callback?: any) => any;
  getSchema: (path?: string) => any;
  reset: (forceClear?: boolean) => void;
  submit: () => Promise<any>;
  validate: () => Promise<any>;
  dispatch: (type: string, payload?: any) => void;
};

export const createFormActions: () => FormActionType;

declare type FormAsyncActionType = {
  getFormState: (callback?: any) => Promise<any>;
  getFieldState: (path?: string, callback: any) => Promise<any>;
  setFormState: (callback?: any) => Promise<any>;
  setFieldState: (path?: string, buffer?: any, callback?: any) => Promise<any>;
  getSchema: (path?: string) => Promise<any>;
  reset: (forceClear?: boolean) => Promise<any>;
  submit: () => Promise<any>;
  validate: () => Promise<any>;
  dispatch: (type: string, payload?: any) => Promise<any>;
};

export const createAsyncFormActions: () => FormAsyncActionType;

export const registerFormField: (type: string, component: any) => void;

export const registerFormFields: (description: RegistedFormFieldsDescription) => void;

export const connect: (options: FieldConnectOptions) => (component: any) => any;