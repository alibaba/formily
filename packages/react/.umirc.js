import { resolve } from 'path'
export default {
  mode: 'site',
  logo: '//img.alicdn.com/imgextra/i2/O1CN01Kq3OHU1fph6LGqjIz_!!6000000004056-55-tps-1141-150.svg',
  title: 'React',
  hash: true,
  favicon:
    '//img.alicdn.com/imgextra/i3/O1CN01XtT3Tv1Wd1b5hNVKy_!!6000000002810-55-tps-360-360.svg',
  outputPath: './doc-site',
  navs: {
    'en-US': [
      {
        title: 'Guide',
        path: '/guide',
      },
      {
        title: 'API',
        path: '/api',
      },
      {
        title: 'Home Site',
        path: 'https://formilyjs.org',
      },
      {
        title: 'GITHUB',
        path: 'https://github.com/alibaba/formily',
      },
    ],
    'zh-CN': [
      {
        title: '指南',
        path: '/zh-CN/guide',
      },
      {
        title: 'API',
        path: '/zh-CN/api',
      },
      {
        title: '主站',
        path: 'https://formilyjs.org',
      },
      {
        title: 'GITHUB',
        path: 'https://github.com/alibaba/formily',
      },
    ],
  },
  links: [
    {
      rel: 'stylesheet',
      href: 'https://unpkg.com/antd/dist/antd.css',
    },
  ],
  styles: [
    `.__dumi-default-navbar-logo{
      background-size: 140px!important;
      background-position: center left!important;
      background-repeat: no-repeat!important;
      padding-left: 150px!important;/*可根据title的宽度调整*/
      font-size: 18px!important;
    }
    .__dumi-default-navbar{
      padding: 0 28px !important;
    }
    .__dumi-default-layout-hero{
      background-image: url(//img.alicdn.com/imgextra/i4/O1CN01ZcvS4e26XMsdsCkf9_!!6000000007671-2-tps-6001-4001.png);
      background-size: cover;
      background-repeat: no-repeat;
      padding: 120px 0 !important;
    }
    .__dumi-default-layout-hero h1{
      color:#45124e !important;
      font-size:80px !important;
      padding-bottom: 30px !important;
    }
    .__dumi-default-dark-switch {
      display:none
    }
    nav a{
      text-decoration: none !important;
    }
    `,
  ],
  menus: {
    '/guide': [
      {
        title: 'Introduction',
        path: '/guide',
      },
      { title: 'Architecture', path: '/guide/architecture' },
      { title: 'Concept', path: '/guide/concept' },
    ],
    '/zh-CN/guide': [
      {
        title: '介绍',
        path: '/guide',
      },
      { title: '核心架构', path: '/zh-CN/guide/architecture' },
      { title: '核心概念', path: '/zh-CN/guide/concept' },
    ],
  },
}
