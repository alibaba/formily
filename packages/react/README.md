# @uform/react

> UForm React Pure Package


### ArrayStringList

```jsx
import React, { useState } from 'react'
import {
  Form,
  Field,
  FormPath,
  createFormActions,
  FormSpy,
  FormProvider,
  FormConsumer,
  FormEffectHooks
} from './src'

const actions = createFormActions()

const Input = props => (
  <Field {...props}>
    {({ state, mutators }) => (
      <div>
        <input
          disabled={!state.editable}
          value={state.value || ''}
          onChange={mutators.change}
          onBlur={mutators.blur}
          onFocus={mutators.focus}
        />
        {state.errors}
        {state.warnings}
      </div>
    )}
  </Field>
)

const { onFormInit$, onFormInputChange$, onFieldInputChange$ } = FormEffectHooks

const App = () => {
  const [values, setValues] = useState({})
  const [editable, setEditable] = useState(true)
  return (
    <FormProvider>
      <Form
        //actions={actions}
        editable={editable}
        initialValues={values}
        effects={() => {
          onFormInit$().subscribe(() => {
            console.log('初始化')
          })
          onFieldInputChange$().subscribe(state => {
            console.log('输入变化', state)
          })
        }}
        onChange={() => {}}
      >
        <Field name="array" initialValue={[]}>
          {({ state, mutators }) => {
            return (
              <div>
                {state.value.map((item, index) => {
                  return (
                    <React.Fragment key={index}>
                      <Input
                        name={`array[${index}]`}
                        required
                        triggerType="onBlur"
                      />
                      <button
                        onClick={() => {
                          mutators.remove(index)
                        }}
                      >
                        Remove
                      </button>
                    </React.Fragment>
                  )
                })}
                <button onClick={() => mutators.push()}>Add Item</button>
              </div>
            )
          }}
        </Field>
      </Form>
    </FormProvider>
  )
}

ReactDOM.render(<App />, document.getElementById('root'))
```

### ArrayObjectList

```jsx
import React, { useState } from 'react'
import {
  Form,
  Field,
  FormPath,
  createFormActions,
  FormSpy,
  FormProvider,
  FormConsumer,
  FormEffectHooks
} from './src'

const actions = createFormActions()

const Input = props => (
  <Field {...props}>
    {({ state, mutators }) => (
      <div>
        <input
          disabled={!state.editable}
          value={state.value || ''}
          onChange={mutators.change}
          onBlur={mutators.blur}
          onFocus={mutators.focus}
        />
        {state.errors}
        {state.warnings}
      </div>
    )}
  </Field>
)

const { onFormInit$, onFormInputChange$, onFieldInputChange$ } = FormEffectHooks

const App = () => {
  const [values, setValues] = useState({})
  const [editable, setEditable] = useState(true)
  return (
    <FormProvider>
      <Form
        //actions={actions}
        editable={editable}
        initialValues={values}
        effects={() => {
          onFormInit$().subscribe(() => {
            console.log('初始化')
          })
          onFieldInputChange$().subscribe(state => {
            console.log('输入变化', state)
          })
        }}
        onChange={() => {}}
      >
        <Field name="array" initialValue={[]}>
          {({ state, mutators }) => {
            return (
              <div>
                {state.value.map((item, index) => {
                  return (
                    <React.Fragment key={index}>
                      <Input
                        name={`array[${index}].aaa`}
                        required
                        triggerType="onBlur"
                      />
                      <Input
                        name={`array[${index}].bbb`}
                        rules={value => {
                          if (value == '123') {
                            return {
                              type: 'warning',
                              message: '这个是一个提示'
                            }
                          }
                        }}
                      />
                      <button
                        onClick={() => {
                          mutators.remove(index)
                        }}
                      >
                        Remove
                      </button>
                    </React.Fragment>
                  )
                })}
                <button onClick={() => mutators.push()}>Add Item</button>
              </div>
            )
          }}
        </Field>
      </Form>
    </FormProvider>
  )
}

ReactDOM.render(<App />, document.getElementById('root'))
```

### Dynamic Object

```jsx
import React, { useState } from 'react'
import {
  Form,
  Field,
  FormPath,
  createFormActions,
  FormSpy,
  FormProvider,
  FormConsumer,
  FormEffectHooks
} from './src'

const actions = createFormActions()

const Input = props => (
  <Field {...props}>
    {({ state, mutators }) => (
      <div>
        <input
          disabled={!state.editable}
          value={state.value || ''}
          onChange={mutators.change}
          onBlur={mutators.blur}
          onFocus={mutators.focus}
        />
        {state.errors}
        {state.warnings}
      </div>
    )}
  </Field>
)

const { onFormInit$, onFormInputChange$, onFieldInputChange$ } = FormEffectHooks

const App = () => {
  const [values, setValues] = useState({})
  const [editable, setEditable] = useState(true)
  return (
    <FormProvider>
      <Form
        actions={actions}
        editable={editable}
        initialValues={values}
        effects={() => {
          onFormInit$().subscribe(() => {
            console.log('初始化')
          })
          onFieldInputChange$().subscribe(state => {
            console.log('输入变化', state)
          })
        }}
        onChange={() => {}}
      >
        <Field name="object" initialValue={{}}>
          {({ state, mutators }) => {
            return (
              <React.Fragment>
                {Object.keys(state.value).map(key => {
                  if (!mutators.exist(key)) return
                  return (
                    <React.Fragment key={key}>
                      <Input
                        name={`object.${key}`}
                        required
                        triggerType="onBlur"
                      />
                      <button
                        onClick={() => {
                          mutators.remove(key)
                          console.log(actions.getFormGraph())
                        }}
                      >
                        Remove
                      </button>
                    </React.Fragment>
                  )
                })}
                <button
                  onClick={() => {
                    mutators.change({
                      ...state.value,
                      [new Date().getTime()]: new Date().getTime()
                    })
                  }}
                >
                  add
                </button>
              </React.Fragment>
            )
          }}
        </Field>
      </Form>
    </FormProvider>
  )
}

ReactDOM.render(<App />, document.getElementById('root'))
```
