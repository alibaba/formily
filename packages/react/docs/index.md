---
title: Formily-Alibaba unified front-end form solution
order: 10
hero:
  title: FORMILY REACT
  desc: Alibaba unified front-end form solution
  actions:
    - text: Home Site
      link: //v2.formilyjs.org
    - text: Development Guide
      link: /guide
features:
  - icon: https://img.alicdn.com/imgextra/i1/O1CN01bHdrZJ1rEOESvXEi5_!!6000000005599-55-tps-800-800.svg
    title: Ultra High Performance
    desc: Dependency tracking, efficient update, on-demand rendering
  - icon: https://img.alicdn.com/imgextra/i2/O1CN016i72sH1c5wh1kyy9U_!!6000000003550-55-tps-800-800.svg
    title: Out Of The Box
    desc: The component status is automatically bound, and the access cost is extremely low
  - icon: https://img.alicdn.com/imgextra/i3/O1CN01JHzg8U1FZV5Mvt012_!!6000000000501-55-tps-800-800.svg
    title: JSON Schema Driver
    desc: Standard JSON-Schema
  - icon: https://img.alicdn.com/imgextra/i3/O1CN0194OqFF1ui6mMT4g7O_!!6000000006070-55-tps-800-800.svg
    title: Scene Reuse
    desc: Based on protocol-driven, abstract scene components
  - icon: https://img.alicdn.com/imgextra/i4/O1CN018vDmpl2186xdLu6KI_!!6000000006939-55-tps-800-800.svg
    title: Debugging Friendly
    desc: Natural docking with Formily DevTools
  - icon: https://img.alicdn.com/imgextra/i4/O1CN01u6jHgs1ZMwXpjAYnh_!!6000000003181-55-tps-800-800.svg
    title: Smart Tips
    desc: Embrace Typescript
footer: Open-source MIT Licensed | Copyright Â© 2019-present<br />Powered by self
---

## Installation

```bash
$ npm install --save @formily/core @formily/react

```

## Quick start

```tsx
/**
 * defaultShowCode: true
 */
import React, { useMemo } from 'react'
import { createForm, setValidateLanguage } from '@formily/core'
import {
  FormProvider,
  FormConsumer,
  Field,
  useField,
  observer,
} from '@formily/react'
import { Input, Form } from 'antd'

// FormItem UI component
const FormItem = observer(({ children }) => {
  const field = useField()
  return (
    <Form.Item
      label={field.title}
      help={field.errors?.length ? field.errors : undefined}
      extra={field.description}
      validateStatus={field.validateStatus}
    >
      {children}
    </Form.Item>
  )
})

/*
 * The above logic has been implemented in @formily/antd, and there is no need to rewrite it in actual use
 */

//Switch the built-in check internationalization copy to English
setValidateLanguage('en')

export default () => {
  const form = useMemo(() => createForm({ validateFirst: true }))

  const createPasswordEqualValidate = (equalName) => (field) => {
    if (
      form.values.confirm_password &&
      field.value &&
      form.values[equalName] !== field.value
    ) {
      field.errors = ['Password does not match Confirm Password.']
    } else {
      field.errors = []
    }
  }

  return (
    <FormProvider form={form}>
      <Form layout="vertical">
        <Field
          name="name"
          title="Name"
          required
          decorator={[FormItem]}
          component={[Input, { placeholder: 'Please Input' }]}
        />
        <Field
          name="password"
          title="Password"
          required
          decorator={[FormItem]}
          component={[Input, { type: 'password', placeholder: 'Please Input' }]}
          reactions={createPasswordEqualValidate('confirm_password')}
        />
        <Field
          name="confirm_password"
          title="Confirm Password"
          required
          decorator={[FormItem]}
          component={[Input, { type: 'password', placeholder: 'Please Input' }]}
          reactions={createPasswordEqualValidate('password')}
        />
        <code>
          <pre>
            <FormConsumer>
              {(form) => JSON.stringify(form.values, null, 2)}
            </FormConsumer>
          </pre>
        </code>
      </Form>
    </FormProvider>
  )
}
```
