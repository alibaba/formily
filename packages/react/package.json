{
  "name": "@formily/react",
  "version": "2.0.0-beta.83",
  "license": "MIT",
  "main": "lib",
  "module": "esm",
  "umd:main": "dist/formily.react.umd.production.js",
  "unpkg": "dist/formily.react.umd.production.js",
  "jsdelivr": "dist/formily.react.umd.production.js",
  "jsnext:main": "esm",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/alibaba/formily.git"
  },
  "types": "esm/index.d.ts",
  "bugs": {
    "url": "https://github.com/alibaba/formily/issues"
  },
  "homepage": "https://github.com/alibaba/formily#readme",
  "engines": {
    "npm": ">=3.0.0"
  },
  "scripts": {
    "start": "dumi dev",
    "build": "rimraf -rf lib esm dist && npm run build:cjs && npm run build:esm && npm run build:umd && npm run build:global",
    "build:cjs": "tsc --declaration",
    "build:esm": "tsc --declaration --module es2015 --outDir esm",
    "build:umd": "rollup --config",
    "build:global": "ts-node ../../scripts/build-global",
    "build:docs": "dumi build"
  },
  "peerDependencies": {
    "@types/react": ">=16.8.0 || >=17.0.0",
    "@types/react-dom": ">=16.8.0 || >=17.0.0",
    "react": ">=16.8.0 || >=17.0.0",
    "react-dom": ">=16.8.0 || >=17.0.0",
    "react-is": ">=16.8.0 || >=17.0.0"
  },
  "devDependencies": {
    "dumi": "^1.1.0-rc.8"
  },
  "dependencies": {
    "@formily/core": "2.0.0-beta.83",
    "@formily/json-schema": "2.0.0-beta.83",
    "@formily/reactive": "2.0.0-beta.83",
    "@formily/reactive-react": "2.0.0-beta.83",
    "@formily/shared": "2.0.0-beta.83",
    "@formily/validator": "2.0.0-beta.83",
    "hoist-non-react-statics": "^3.3.2"
  },
  "publishConfig": {
    "access": "public"
  },
  "gitHead": "ac79c196ae9324889aca5e0501146f9b37b04283"
}
