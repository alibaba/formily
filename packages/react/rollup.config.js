import baseConfig from '../../scripts/rollup.base.js'

export default baseConfig('formily.react', 'Formily.React')
