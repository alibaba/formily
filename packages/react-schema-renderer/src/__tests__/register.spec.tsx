import React from 'react'
import {
  registerFormField,
  registerFieldMiddleware,
  registerVirtualBox,
  connect,
  SchemaForm,
  createFormActions,
  cleanRegistry
} from '../index'
import { render, fireEvent, wait } from '@testing-library/react'

describe('test all apis', () => {
  afterEach(() => {
    cleanRegistry()
  })
  test('registerFormField', async () => {
    registerFormField(
      'string',
      connect()(props => {
        return <input {...props} value={props.value || ''} />
      })
    )

    const actions = createFormActions()

    const { queryByTestId } = render(
      <SchemaForm
        actions={actions}
        schema={{
          type: 'object',
          properties: {
            aa: {
              type: 'string',
              default: '123',
              'x-props': {
                'data-testid': 'input'
              }
            }
          }
        }}
      />
    )
    expect(queryByTestId('input').getAttribute('value')).toEqual('123')
    fireEvent.change(queryByTestId('input'), {
      target: {
        value: 'aa change'
      }
    })
    await wait(() => {
      expect(queryByTestId('input').getAttribute('value')).toEqual('aa change')
      expect(actions.getFormGraph()).toMatchSnapshot()
    })
  })

  test('registerFieldMiddleware', async () => {
    registerFormField(
      'string',
      connect()(props => {
        return <input {...props} value={props.value || ''} />
      })
    )

    registerFieldMiddleware(FieldComponent => props => {
      return (
        <div>
          this is wrapper
          <FieldComponent {...props} />
        </div>
      )
    })

    const actions = createFormActions()

    const { queryByTestId, queryByText } = render(
      <SchemaForm
        actions={actions}
        schema={{
          type: 'object',
          properties: {
            aa: {
              type: 'string',
              default: '123',
              'x-props': {
                'data-testid': 'input'
              }
            }
          }
        }}
      />
    )
    expect(queryByTestId('input').getAttribute('value')).toEqual('123')
    fireEvent.change(queryByTestId('input'), {
      target: {
        value: 'aa change'
      }
    })
    await wait(() => {
      expect(queryByTestId('input').getAttribute('value')).toEqual('aa change')
      expect(actions.getFormGraph()).toMatchSnapshot()
      expect(queryByText('this is wrapper')).toBeTruthy()
    })
  })

  test('registerVirtualBox', async () => {
    registerFormField(
      'string',
      connect()(props => {
        return <input {...props} value={props.value || ''} />
      })
    )

    registerVirtualBox('box', props => {
      return <div>this is VirtualBox.{props.children}</div>
    })

    const actions = createFormActions()

    const { queryByTestId, queryByText } = render(
      <SchemaForm
        actions={actions}
        schema={{
          type: 'object',
          properties: {
            cc: {
              type: 'object',
              'x-component': 'box',
              properties: {
                aa: {
                  type: 'string',
                  default: '123',
                  'x-props': {
                    'data-testid': 'input'
                  }
                }
              }
            }
          }
        }}
      />
    )
    expect(queryByTestId('input').getAttribute('value')).toEqual('123')
    fireEvent.change(queryByTestId('input'), {
      target: {
        value: 'aa change'
      }
    })
    await wait(() => {
      expect(queryByTestId('input').getAttribute('value')).toEqual('aa change')
      expect(actions.getFormGraph()).toMatchSnapshot()
      expect(actions.getFormState(state => state.values)).toEqual({
        aa: 'aa change'
      })
      expect(queryByText('this is VirtualBox.')).toBeTruthy()
    })
  })
})

describe('major scenes',()=>{
  //todo
  test('basic',()=>{
    //todo
  })
})


describe('bugfix',()=>{
  //todo
  test('basic',()=>{
    //todo
  })
})