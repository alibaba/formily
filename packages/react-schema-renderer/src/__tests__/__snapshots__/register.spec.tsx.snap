// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test all apis registerFieldMiddleware 1`] = `
Object {
  "": Object {
    "displayName": "FormState",
    "editable": undefined,
    "errors": Array [],
    "initialValues": Object {
      "aa": "123",
    },
    "initialized": true,
    "invalid": false,
    "loading": false,
    "modified": true,
    "mounted": true,
    "props": Object {},
    "submitting": false,
    "unmounted": false,
    "valid": true,
    "validating": false,
    "values": Object {
      "aa": "aa change",
    },
    "warnings": Array [],
  },
  "aa": Object {
    "active": false,
    "dataType": "string",
    "display": true,
    "displayName": "FieldState",
    "editable": true,
    "effectErrors": Array [],
    "effectWarnings": Array [],
    "errors": Array [],
    "formEditable": undefined,
    "initialValue": "123",
    "initialized": true,
    "invalid": false,
    "loading": false,
    "modified": true,
    "mounted": true,
    "name": "aa",
    "path": "aa",
    "pristine": false,
    "props": Object {
      "default": "123",
      "key": "aa",
      "type": "string",
      "x-props": Object {
        "data-testid": "input",
      },
    },
    "required": false,
    "ruleErrors": Array [],
    "ruleWarnings": Array [],
    "rules": Array [],
    "selfEditable": undefined,
    "touched": false,
    "unmountRemoveValue": true,
    "unmounted": false,
    "valid": true,
    "validating": false,
    "value": "aa change",
    "values": Array [
      "aa change",
    ],
    "visible": true,
    "visibleCacheValue": undefined,
    "visited": false,
    "warnings": Array [],
  },
}
`;

exports[`test all apis registerFormField 1`] = `
Object {
  "": Object {
    "displayName": "FormState",
    "editable": undefined,
    "errors": Array [],
    "initialValues": Object {
      "aa": "123",
    },
    "initialized": true,
    "invalid": false,
    "loading": false,
    "modified": true,
    "mounted": true,
    "props": Object {},
    "submitting": false,
    "unmounted": false,
    "valid": true,
    "validating": false,
    "values": Object {
      "aa": "aa change",
    },
    "warnings": Array [],
  },
  "aa": Object {
    "active": false,
    "dataType": "string",
    "display": true,
    "displayName": "FieldState",
    "editable": true,
    "effectErrors": Array [],
    "effectWarnings": Array [],
    "errors": Array [],
    "formEditable": undefined,
    "initialValue": "123",
    "initialized": true,
    "invalid": false,
    "loading": false,
    "modified": true,
    "mounted": true,
    "name": "aa",
    "path": "aa",
    "pristine": false,
    "props": Object {
      "default": "123",
      "key": "aa",
      "type": "string",
      "x-props": Object {
        "data-testid": "input",
      },
    },
    "required": false,
    "ruleErrors": Array [],
    "ruleWarnings": Array [],
    "rules": Array [],
    "selfEditable": undefined,
    "touched": false,
    "unmountRemoveValue": true,
    "unmounted": false,
    "valid": true,
    "validating": false,
    "value": "aa change",
    "values": Array [
      "aa change",
    ],
    "visible": true,
    "visibleCacheValue": undefined,
    "visited": false,
    "warnings": Array [],
  },
}
`;

exports[`test all apis registerVirtualBox 1`] = `
Object {
  "": Object {
    "displayName": "FormState",
    "editable": undefined,
    "errors": Array [],
    "initialValues": Object {
      "aa": "123",
    },
    "initialized": true,
    "invalid": false,
    "loading": false,
    "modified": true,
    "mounted": true,
    "props": Object {},
    "submitting": false,
    "unmounted": false,
    "valid": true,
    "validating": false,
    "values": Object {
      "aa": "aa change",
    },
    "warnings": Array [],
  },
  "cc": Object {
    "display": true,
    "displayName": "VirtualFieldState",
    "initialized": true,
    "mounted": true,
    "name": "cc",
    "path": "cc",
    "props": Object {
      "key": "cc",
      "type": "object",
      "x-component": "box",
    },
    "unmounted": false,
    "visible": true,
  },
  "cc.aa": Object {
    "active": false,
    "dataType": "string",
    "display": true,
    "displayName": "FieldState",
    "editable": true,
    "effectErrors": Array [],
    "effectWarnings": Array [],
    "errors": Array [],
    "formEditable": undefined,
    "initialValue": "123",
    "initialized": true,
    "invalid": false,
    "loading": false,
    "modified": true,
    "mounted": true,
    "name": "aa",
    "path": "cc.aa",
    "pristine": false,
    "props": Object {
      "default": "123",
      "key": "aa",
      "type": "string",
      "x-props": Object {
        "data-testid": "input",
      },
    },
    "required": false,
    "ruleErrors": Array [],
    "ruleWarnings": Array [],
    "rules": Array [],
    "selfEditable": undefined,
    "touched": false,
    "unmountRemoveValue": true,
    "unmounted": false,
    "valid": true,
    "validating": false,
    "value": "aa change",
    "values": Array [
      "aa change",
    ],
    "visible": true,
    "visibleCacheValue": undefined,
    "visited": false,
    "warnings": Array [],
  },
}
`;
